Verification links with expiration time

johanndev1208 opened the pull request:
**Here's a quick summary of some points to check**
Added expiry for email verification links (valid for 72 hours only)

## References

According [doc](https://github.com/heartcombo/devise/blob/cf93de390a29654620fdf7ac07b4794eb95171d0/lib/devise/models/recoverable.rb#L90C16-L90C19) currently token is used in invalid way. 

## Objectives

- Added expiry for email verification links (valid for 72 hours only)
- Aligns with GDPR principles of data minimization (Art. 5(1)(c))
  and storage limitation (Art. 5(1)(e))
- Improves overall account security (Art. 32 GDPR)

johanndev1208 commented on app/models/verification/email.rb:
```diff
@@ -23,7 +23,7 @@ def user
 
   def generate_token
-    @plain_token, @encrypted_token = Devise.token_generator.generate(User, :email_verification_token)
+    @encrypted_token, @plain_token = Devise.token_generator.generate(User, :email_verification_token)
```
https://github.com/heartcombo/devise/blob/cf93de390a29654620fdf7ac07b4794eb95171d0/lib/devise/models/recoverable.rb#L90C16-L90C19
in this resourse we can find what should we store to db and use in payload.

javierm commented on app/models/verification/management/email.rb:
```diff
@@ -21,7 +21,7 @@ def user?
     return false unless valid?
 
-    plain_token, encrypted_token = Devise.token_generator.generate(User, :email_verification_token)
+    encrypted_token, plain_token = Devise.token_generator.generate(User, :email_verification_token)
```
Same here; instead of changing the variable names, we could use `email_verification_token: encrypted_token` in the `user.update!` call.

javierm reviewed (commented):
@johanndev1208 Thanks for the pull request! :pray:

For now, I'd vote to simply uncomment the `config.confirm_within = 3.days` and write the test for that :ok_hand:.

We can remove the rest of the changes for now, but if you think they're necessary, you can open a different pull request with an explanation about how they would affect existing Consul Democracy installations :thinking:.

taitus committed:
Expire email confirmation links after 3 days

Enable `config.confirm_within = 3.days` in Devise so that
confirmation links become invalid 72 hours after being sent.

Why:

- Old confirmation links left in email inboxes, backups or logs
  should not remain valid forever.
- Aligns with GDPR data minimization and storage limitation
  (Art. 5(1)(c) and 5(1)(e)) and account security (Art. 32).

Why it is not just one line in devise.rb:

The confirmations controller has a custom flow for management
users who do not have a password yet. In that flow the
controller saves the password first and then calls
`resource.confirm`.

With `confirm_within` enabled, an expired link would still
render the password form and save the password, but fail to
confirm the user silently. The user would then end up on the
login screen with no hint that the link had expired, trying
to sign in with a password that does not work.

To avoid that, the controller now rejects expired links
before showing any password form:

- `show` returns early with `super`. Devise then renders its
  default "resend confirmation instructions" screen with the
  standard expiration error.
- `update` redirects to `show` when the link is expired, so
  the same response is reused and the code is not duplicated.

taitus committed:
Use set_flash_message! in confirmations controller

`set_flash_message!` was added in Devise 4.6 and already checks
`is_flashing_format?` internally, so the explicit guard is
redundant.

Behavior is identical: the flash is still only set on
navigational formats that respond to flash.

javierm committed:
Simplify logic in confirmations controller

Instead of copying the logic from the Devise controller and changing it,
we can call `super` except for the case where the encrypted password is
blank. This case is what caused us to customize this controller.

Note that since commit 0195eac3f we were responding with a 422 status
code when the token had already been used. However, Devise only responds
with the 422 status code on non-GET requests. For GET requests, it
responds with the status 200. So we're updating the test that checked
that.

javierm commented on app/controllers/users/confirmations_controller.rb:
```diff
@@ -81,4 +67,8 @@ def allowed_params
     end
+
+    def confirmation_period_expired?
+      resource.present? && resource.send(:confirmation_period_expired?)
```
Do we need `resource.present?` here? 🤔 In the `show` action, we use `find_by!`, so this shouldn't be needed. In the `update` action we do use `find_by`, but this is probably something we forgot to change in commit 78c6395e5f4 🤔. The `update` action currently generates a 500 error when the resource isn't found.

So what do you think about using `find_by!` in the `update` action and removing the `resource.present? &&` part of this line?

javierm commented on spec/system/management/users_spec.rb:
```diff
@@ -44,6 +44,24 @@
+      expect(page).to have_content "You need to be verified within 3 days"
+      expect(page).not_to have_field "New access password"
+      expect(user.reload.encrypted_password).to be_blank
+      expect(user.reload).not_to be_confirmed
```
Here we're checking the database after starting the browser. Shall we remove these lines? We're already checking this in the controller tests after all 🤔.

taitus committed:
Use find_by! in confirmations#update

In the show action we use find_by!, but in the update action we
still use find_by. This is probably something we forgot to change
in commit 78c6395, where show was switched to find_by! to avoid
returning a 500 Internal Server Error on invalid tokens.

The update action currently generates a 500 error when the
resource isn't found, since the code that follows assumes
find_by returned an object responding to encrypted_password.

Use find_by! here as well so an invalid token in update responds
with 404 instead of 500, consistently with show.

Note that we align the existing show test to pass the token under the
`confirmation_token` key, matching what the controller reads and the
new update test.
